composer/datasets/in_context_learning_evaluation.py

```diff
@@ -100,9 +142,168 @@ def update_metric(self, metric, batch, output_logits, labels):
+        tokenizer: Union[transformers.PreTrainedTokenizer, transformers.PreTrainedTokenizerFast],
+        max_seq_len: int,
+        eos_tok_id: int,
+        num_fewshot: int,
+        preamble_string: str,
+        example_delimiter: str,
+        continuation_delimiter: str,
+        destination_path: str = 'icl_mc_task.json',
+    ):
+        get_file(dataset_uri, destination_path, overwrite=True)
+        dataset = load_dataset('json', data_files=destination_path, split='train', streaming=False)
+        self.samples = list(
+            dataset.map(lambda examples: {
+                'query': examples['query'],
+                'choices': examples['choices'],
+                'gold': examples['gold']
+            }))
+        self.num_choices = len(self.samples[0]['choices'])
+        self.tokenizer = tokenizer
+        self.max_seq_len = max_seq_len
+        self.eos_tok_id = eos_tok_id
+        self.encoded_dataset = self.prep_examples(num_fewshot, preamble_string, example_delimiter,
+                                                  continuation_delimiter)
+
+    def prep_examples(self, num_fewshot, preamble_string, example_delimiter, continuation_delimiter):
+        examples = []
+        for sample_idx in range(len(self.samples)):
+
+            preamble = preamble_string
+            if num_fewshot > 0:
+                allowable_indices = list(range(len(self.samples)))
+                allowable_indices.remove(sample_idx)
+                fewshot_idxs = random.sample(allowable_indices, num_fewshot)
+
+                for fewshot_idx in fewshot_idxs:
+                    query, choices, gold_idx = self.samples[fewshot_idx]['query'], self.samples[fewshot_idx][
+                        'choices'], self.samples[fewshot_idx]['gold']
+                    if len(preamble) > 0:
+                        query = f'{example_delimiter}{query}'
+                    preamble += f'{query}{continuation_delimiter}{choices[gold_idx]}'
+
+            encoded_example = {}
+            query, choices, gold_idx = self.samples[sample_idx]['query'], self.samples[sample_idx][
+                'choices'], self.samples[sample_idx]['gold'],
+            if len(preamble) > 0:
+                query = f'{example_delimiter}{query}'
+
+            choices = [f'{continuation_delimiter}{choice}' for choice in choices]
+            encoded_example['query'] = self.tokenizer(query)
+            encoded_example['choices'] = [self.tokenizer(choice) for choice in choices]
+            encoded_example['preamble'] = self.tokenizer(preamble)
+            encoded_example['gold_idx'] = gold_idx
+
+            examples.append(encoded_example)
+
+        return examples
+
+    def __getitem__(self, index):
+        return self.encoded_dataset[index]
+
+    def __len__(self):
+        return len(self.encoded_dataset)
+
+    def collate_fn(self, data):
+        inputs = []
+        continuation_indices = []
+        gold_idxs = []
+        choice_groupings = []
```

Comment:
could you comment explaining the `choice_groupings` list? That one in particular took me a bit to understand that it was indexing into `continuation_indices` because each choice is its own continuation.